ramips: add support for D-Link DIR-882 R1

pyckle opened the pull request:
The R1 revision is identical to the A1 revision except
- No Config2 Parition, therefore
   - factory partition resized to 64k from 128K
   - Firmware partition offset is 0x50000 not 0x60000
   - Firmware partitions size increased by 64K
- Firmware partition type is "denx,uimage", not "sge,uimage"
- Padding of image creation "uimage-padhdr 96" removed

Installation:
- Older firmware versions: put the factory image on a USB stick, turn on
  the telnet console, and flash using the following cmd
  "fw_updater Linux /mnt/usb_X_X/firmware.bin"

- D-Link FailsafeUI:
  Sign the firmware image with the script here:
  https://pastebin.com/5wGYjTsn
  Then, power down the router, press and hold the reset button, then
  re-plug it. Keep the reset button pressed until the internet LED stops
  flashing, then jack into any lan port and manually assign a static IP
  address in 192.168.0.0/24 other than 192.168.0.0 (e.g. 192.168.0.2)
  and go to http://192.168.0.1

Signed-off-by: Andrew Pikler <andrew.pikler@gmail.com>

adschm commented on target/linux/ramips/dts/mt7621_dlink_dir-882-r1.dts:
```diff
@@ -0,0 +1,54 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+/dts-v1/;
```
drop dts-v1, it's in mt7621.dtsi now ...

pyckle replied:
Done.

adschm commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -250,6 +250,18 @@ define Device/dlink_dir-8xx-a1
+  IMAGE/sysupgrade.bin := append-kernel | append-rootfs |\
+	pad-rootfs | append-metadata | check-size
+  IMAGE/factory.bin := append-kernel | append-rootfs | check-size
+endef
```
Do you expect other similar devices? If not, just put the command into the device's node as long as only one device inherits from it.

pyckle replied:
I think that the DIR-878-R1 will have the same partition structure and can reuse these files.

adschm commented:
Please add specifications to your commit message.

Despite, you are settings execute bit on all files, please remove that unless where necessary (01_leds).

pyckle commented:
Hi,

Thanks for the timely review (:

I added the specifications to the commit message, removed the execute bits that mistakenly were added, and commented/changed on the other feedback messages.

I am also wondering whether it is worth getting the code to sign the factory image so it can be more easily uploaded onto the router into the build process. Currently it is a bash script here: https://pastebin.com/5wGYjTsn

If I were to do this, where should I place the tool and what language should it be in? Can rewrite this in bash, C, python or perl.

pyckle commented:
I added the signing process as part of the build. Please let me know what you think.

adschm commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -317,9 +328,20 @@ define Device/dlink_dir-882-a1
   DEVICE_PACKAGES += kmod-usb3 kmod-usb-ledtrig-usbport
+  IMAGE/factory.bin := append-kernel | append-rootfs | check-size | \
+	  sign-dlink-ru 57c5375741c30ca9ebcb36713db4ba51 \
+	  ab0dff19af8842cdb70a86b4b68d23f7
```
That's added to the wrong device?

Anyway, I'd make the factory image stuff a separate commit ...

pyckle replied:
Whoops! Moved to the right section.

I moved the factory image stuff to a separate commit now too.

adschm commented on target/linux/ramips/dts/mt7621_dlink_dir-882-r1.dts:
```diff
@@ -0,0 +1,53 @@
+		compatible = "gpio-leds";
+
+		led_power_orange: power_orange {
+			label = "dir-882-r1:orange:power";
```
I've just pushed a patch to master that removes the model part of these LED labels. With that change, keys and leds are the same as for dir-8xx-a1 as it appears to me. So, the question arises whether we can move more to the DTSI or even create a shared DTSI for dir-8xx-a1 and dir-8xx-r1 ...

pyckle replied:
Nice work (: I adjusted the DTSI/DTS changes to reflect your patch. At this point, the dtsi file between the 8xx-r1 routers and 8xx-a1 routers is related to the partitioning.

I think there are a few options here:
1. Keep things as I pushed them.
2. Make an 8xx-x1 dtsi, have the 8xx-r1.dtsi and 8xx-a1.dtsi import it, and put the partition section there.
3. Only have an 8xx-x1 dtsi and put the partition section in each model's dts file.

What would you like me to do? I would appreciate if somebody throughly review any changes if I refactor the DTSI files. I only have the DIR882-R1 model to test with, and my knowledge of DTSI and DTS files is rudimentary.

pyckle replied:
I created a few dtsi files to reduce duplication. Please let me know what you think.

pyckle committed:
firmware: add tool for signing d-link ru router factory firmware images

Some Russian d-link routers require that their firmware be signed with a
salted md5 checksum followed by the bytes 0x00 0xc0 0xff 0xee. This tool
signs factory images the OEM's firmware accepts them.

Signed-off-by: Andrew Pikler <andrew.pikler@gmail.com>

pyckle committed:
ramips: add support for D-Link DIR-882 R1

Specifications:
- SoC: MediaTek MT7621AT
- RAM: 128 MB (DDR3)
- Flash: 16 MB (SPI NOR)
- WiFi: MediaTek MT7615N (x2)
- Switch: 1 WAN, 4 LAN (Gigabit)
- Ports: 1 USB 2.0, 1 USB 3.0
- Buttons: Reset, WiFi Toggle, WPS
- LEDs: Power, Internet, WiFi 2.4G WiFi 5G, USB 2.0, USB 3.0

The R1 revision is identical to the A1 revision except
- No Config2 Parition, therefore
- factory partition resized to 64k from 128K
- Firmware partition offset is 0x50000 not 0x60000
- Firmware partitions size increased by 64K
- Firmware partition type is "denx,uimage", not "sge,uimage"
- Padding of image creation "uimage-padhdr 96" removed

Installation:
- Older firmware versions: put the factory image on a USB stick, turn on
the telnet console, and flash using the following cmd
"fw_updater Linux /mnt/usb_X_X/firmware.bin"

- D-Link FailsafeUI:
Power down the router, press and hold the reset button, then
re-plug it. Keep the reset button pressed until the internet LED stops
flashing, then jack into any lan port and manually assign a static IP
address in 192.168.0.0/24 other than 192.168.0.0 (e.g. 192.168.0.2)
and go to http://192.168.0.1
Flash with the factory image.

Signed-off-by: Andrew Pikler <andrew.pikler@gmail.com>

ayavilevich commented:
@adschm hey,

Do you know what is preventing this PR from being merged? Please advise on next steps. Having an official support for the DIR-882 R1 would be wonderful.

shayx commented:
Hey guys, from where can I download a working image for the dir 882 e1?

ayavilevich commented:
@yehoshuao there is a build at http://luckys.onmypc.net/openwrt/DIR-882-R1/ but I think it is old by now. You can build your own or maybe @pyckle can post an up to date link.

There is also a discussion at https://forum.openwrt.org/t/d-link-dir-882-r1-bootloader-fw-image-verification/74834 . fyi.

pyckle commented:
@yehoshuao 
A build from this morning is available here: https://mrpickles.fr.to/43402406772d41dbfd1e66e064ea7e5c8a3d8f7e7ed382ee8c29b21f7aeb9512/

$ sha256sum bin/targets/ramips/mt7621/openwrt-ramips-mt7621-dlink_dir-882-r1-squashfs-*
5a68c06bc0b64c40d73ca6de3524b57c3c3800850c9283333609072012ab55f8  bin/targets/ramips/mt7621/openwrt-ramips-mt7621-dlink_dir-882-r1-squashfs-factory.bin
43402406772d41dbfd1e66e064ea7e5c8a3d8f7e7ed382ee8c29b21f7aeb9512  bin/targets/ramips/mt7621/openwrt-ramips-mt7621-dlink_dir-882-r1-squashfs-sysupgrade.bin

yeohayon commented:
thanks @ayavilevich :)
and thanks a lot @pyckle for your hard work! will test it soon and report back

ynezz commented:
Thanks! Pulled into my staging tree at https://git.openwrt.org/openwrt/staging/ynezz.git
